### PR TITLE
feat(core): introduce EnvironmentProviders wrapper type

### PR DESCRIPTION
**Pull request by alxhub**
feat(core): introduce EnvironmentProviders wrapper type

This commit introduces a new type `EnvironmentProviders` which can be used
in contexts where Angular accepted `Provider`s destined for
`EnvironmentInjector`s. This includes contexts such as `@NgModule.providers`
and `Route.providers`.

The new type is useful for preventing such providers from accidentally
ending up in `@Component.providers`. It can be used as the return type of
provider functions (such as `provideRouter`) to enforce this safety.

Because `Provider` allows `any[]` nested arrays, the compile-time safety
provided by `EnvironmentProviders` is easily circumvented. However, the
runtime shape of `EnvironmentProviders` is not compatible with component
injectors and will result in a runtime error if it leaks through (NG0207).

A new function `makeEnvironmentProviders` is used to construct this new type
from an array of providers.

The existing `importProvidersFrom` operation previously returned a very
similar type `ImportedNgModuleProviders` which had the same goal. This
machinery is switched over to use the new `EnvironmentProviders` interface
instead (in fact, `ImportedNgModuleProviders` is now just an alias to
`EnvironmentProviders`).

### CODE REVIEW

File: packages/core/src/di/interface/provider.ts [hunk 1]
```diff
@@ -332,6 +332,41 @@ export interface ClassProvider extends ClassSansProvider {
...
+ * @see `makeEnvironmentProviders`
+ * @see `importProvidersFrom`
+ *
+ * @publicApi
+ */
+export type EnvironmentProviders = {
+  ɵbrand: 'EnvironmentProviders';
+};
+
+export interface InternalEnvironmentProviders extends EnvironmentProviders {
```
Comment (AndrewKushnir) [inline packages/core/src/di/interface/provider.ts:354]: nit: may be add a quick comment on why we need this type (I think it's covered in the `ɵfromNgModule` field description, may be just add a version of it as a comment about this function)?

File: packages/core/src/di/provider_collection.ts [hunk 2]
```diff
@@ -21,9 +21,19 @@ import {resolveForwardRef} from './forward_ref';
 import {ENVIRONMENT_INITIALIZER} from './initializer_token';
 import {ɵɵinject as inject} from './injector_compatibility';
 import {getInjectorDef, InjectorType, InjectorTypeWithProviders} from './interface/defs';
-import {ClassProvider, ConstructorProvider, ExistingProvider, FactoryProvider, ImportedNgModuleProviders, ModuleWithProviders, Provider, StaticClassProvider, TypeProvider, ValueProvider} from './interface/provider';
+import {ClassProvider, ConstructorProvider, EnvironmentProviders, ExistingProvider, FactoryProvider, ImportedNgModuleProviders, InternalEnvironmentProviders, isEnvironmentProviders, ModuleWithProviders, Provider, StaticClassProvider, TypeProvider, ValueProvider} from './interface/provider';
 import {INJECTOR_DEF_TYPES} from './internal_tokens';
 
+/**
+ * Wrap an array of `Provider`s into `EnvironmentProviders`, preventing them from being accidentally
+ * referenced in `@Component in a component injector.
```
Comment (AndrewKushnir) [inline packages/core/src/di/provider_collection.ts:29]:
```suggestion
 * Wrap an array of `Provider`s into `EnvironmentProviders`, preventing them from being accidentally
 * referenced in `@Component` or `@Directive` in a component injector.
```

File: packages/core/src/di/provider_collection.ts [hunk 3]
```diff
@@ -251,12 +263,12 @@ export function walkProviderTree(
...
 
     // Next, include providers listed on the definition itself.
-    const defProviders = injDef.providers;
+    const defProviders = injDef.providers as Array<SingleProvider|InternalEnvironmentProviders>;
     if (defProviders != null && !isDuplicate) {
       const injectorType = container as InjectorType<any>;
-      deepForEach(defProviders, provider => {
-        ngDevMode && validateProvider(provider, defProviders as SingleProvider[], injectorType);
-        providersOut.push(provider);
+      deepForEachProvider(defProviders, provider => {
```
Comment (AndrewKushnir) [inline packages/core/src/di/provider_collection.ts:269]: Just curious if we can type it here (vs adding `as`)?
```suggestion
      deepForEachProvider(defProviders, (provider: SingleProvider) => {
```

**Review by AndrewKushnir — APPROVED** [review]
Looks great 👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 07017a7bd30c6bb7bd1b94a9fd1b112ee68c9ced.

**Pull request closed without merging**